tests/vm/tcastint.nim

```diff
@@ -0,0 +1,110 @@
+discard """
+  file: "tcastint.nim"
+  output: "OK"
+"""
+
+type
+  Dollar = distinct int
+  XCoord = distinct int32
+  Digit = range[-9..0]
+
+# those are necessary for comparisons below.
+proc `==`(x, y: Dollar): bool {.borrow.}
+proc `==`(x, y: XCoord): bool {.borrow.}
+
+proc dummy[T](x: T): T = x
+
+proc test() =
+  let U8 = 0b0011_0010'u8
+  let I8 = 0b0011_0010'i8
+  let C8 = 0b0011_0010'u8.char
+  let C8_1 = 0b0011_0011'u8.char
+  let U16 = 0b00100111_00101000'u16
+  let I16 = 0b00100111_00101000'i16
+  let U32 = 0b11010101_10011100_11011010_01010000'u32
+  let I32 = 0b11010101_10011100_11011010_01010000'i32
+  let U64A = 0b01000100_00111111_01111100_10001010_10011001_01001000_01111010_00010001'u64
+  let I64A = 0b01000100_00111111_01111100_10001010_10011001_01001000_01111010_00010001'i64
+  let U64B = 0b00110010_11011101_10001111_00101000_00000000_00000000_00000000_00000000'u64
+  let I64B = 0b00110010_11011101_10001111_00101000_00000000_00000000_00000000_00000000'i64
+  let UX = U64A.uint
+  let IX = I64A.int
+
+  doAssert(cast[char](I8) == C8)
+  doAssert(cast[uint8](I8) == U8)
+  doAssert(cast[uint16](I16) == U16)
+  doAssert(cast[uint32](I32) == U32)
+  doAssert(cast[uint64](I64A) == U64A)
+  doAssert(cast[uint64](I64B) == U64B)
+  doAssert(cast[int8](U8) == I8)
+  doAssert(cast[int16](U16) == I16)
+  doAssert(cast[int32](U32) == I32)
+  doAssert(cast[int64](U64A) == I64A)
+  doAssert(cast[int64](U64B) == I64B)
+  doAssert(cast[uint](IX) == UX) # uint is always 64 bits in VM
+  doAssert(cast[int](UX) == IX) # int is always 64 bits in VM
```

Comment:
This is confirmed by @Araq. int is really always 64-bit in the VM (for the moment anyway). This might change in the future.